core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java

```diff
@@ -244,6 +247,16 @@ public void setUpdateType(String updateType, QualifiedObjectName targetName, Opt
         this.target = Optional.of(new UpdateTarget(targetName, targetTable, targetColumns));
     }
 
+    /*
+     * For EXPLAIN and EXPLAIN ANALYZE queries we would reset the update type thereby
```

Comment:
We should change the analysis to not require reseting the update type. This can be done by adding an indicator that an EXPLAIN or EXPLAIN ANALYZE query is being analyzed and avoid setting the update type altogether during analysis of INSERT/UPDATE/etc altogether when that flag is set.

Reply:
Does this mean that we don't have to update/reset the `target` too or should it happen in a separate API ? By not setting it we might fail for `EXPLAIN` or `EXPLAIN ANALYZE` of queries with `DELETE`\`UPDATE` bcoz we have an additional validation in `BeginTableWrite` optimizer.

Reply:
I think we can extract it as a different PR follow up.. Since we don't rely on update state being set or reset but rather have a introduced a new API. I guess the change would be a bit orthogonal - even after the revamping the analysis we would face the similar issue since `table` won't be updatable.

Reply:
After the latest update, this comment is now inaccurate.

Reply:
Also, this can be further improved by decoupling the setting of the `updateType` and the `target` fields by splitting into two separate methods. In that case, we don't need to keep track of a separate `tableToBeUpdate`, as `target` would always be set for an INSERT/DELETE/UPDATE query

Reply:
But we need it only for `DELETE/UPDATE` query.

Reply:
Instead of splitting into two APIs - as we need target for `CREATE`/`INSERT`/`UPDATE`/`DELETE`. we are setting it for all types of query instead of applying it only for non EXPLAIN type of queries.

Reply:
@martint AC